src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2365,1578 @@ IsForeignTable(Oid relationId)
+	task->owner = DatumGetObjectId(values[Anum_pg_dist_background_task_owner - 1]);
+	if (!nulls[Anum_pg_dist_background_task_pid - 1])
+	{
+		int32 pid = DatumGetInt32(values[Anum_pg_dist_background_task_pid - 1]);
+		SET_NULLABLE_FIELD(task, pid, pid);
+	}
+	task->status = BackgroundTaskStatusByOid(
+		DatumGetObjectId(values[Anum_pg_dist_background_task_status - 1]));
+
+	task->command = text_to_cstring(
+		DatumGetTextP(values[Anum_pg_dist_background_task_command - 1]));
+
+	if (!nulls[Anum_pg_dist_background_task_retry_count - 1])
+	{
+		int32 retryCount =
+			DatumGetInt32(values[Anum_pg_dist_background_task_retry_count - 1]);
+		SET_NULLABLE_FIELD(task, retry_count, retryCount);
+	}
+
+	if (!nulls[Anum_pg_dist_background_task_not_before - 1])
+	{
+		TimestampTz notBefore =
+			DatumGetTimestampTz(values[Anum_pg_dist_background_task_not_before - 1]);
+		SET_NULLABLE_FIELD(task, not_before, notBefore);
+	}
+
+	if (!nulls[Anum_pg_dist_background_task_message - 1])
+	{
+		task->message = pstrdup(
+			DatumGetCString(values[Anum_pg_dist_background_task_message - 1]));
+	}
+
+	return task;
+}
+
+
+/*
+ * BackgroundTaskHasUmnetDependencies checks if a task from the given job has any unmet
+ * dependencies. An unmet dependency is a Task that the task in question depends on and
+ * has not reached its Done state.
+ */
+static bool
+BackgroundTaskHasUmnetDependencies(int64 jobId, int64 taskId)
+{
+	bool hasUnmetDependency = false;
+
+	Relation pgDistBackgroundTasksDepend =
+		table_open(DistBackgroundTaskDependRelationId(), AccessShareLock);
+
+	ScanKeyData scanKey[2] = { 0 };
+	bool indexOK = true;
+
+	/* pg_catalog.pg_dist_background_task_depend.job_id = jobId */
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_background_task_depend_job_id,
+				BTEqualStrategyNumber, F_INT8EQ, jobId);
+
+	/* pg_catalog.pg_dist_background_task_depend.task_id = $taskId */
+	ScanKeyInit(&scanKey[1], Anum_pg_dist_background_task_depend_task_id,
+				BTEqualStrategyNumber, F_INT8EQ, taskId);
+
+	SysScanDesc scanDescriptor =
+		systable_beginscan(pgDistBackgroundTasksDepend,
+						   DistBackgroundTaskDependTaskIdIndexId(),
+						   indexOK, NULL, lengthof(scanKey),
+						   scanKey);
+
+	HeapTuple dependTuple = NULL;
+	while (HeapTupleIsValid(dependTuple = systable_getnext(scanDescriptor)))
+	{
+		Form_pg_dist_background_task_depend depends =
+			(Form_pg_dist_background_task_depend) GETSTRUCT(dependTuple);
+
+		BackgroundTask *dependingJob = GetBackgroundTaskByTaskId(jobId,
+																 depends->depends_on);
+
+		/*
+		 * Only when the status of all depending jobs is done we clear this job and say
+		 * that is has no unmet dependencies.
+		 */
+		if (dependingJob->status == BACKGROUND_TASK_STATUS_DONE)
+		{
+			continue;
+		}
+
+		/*
+		 * we assume that when we ask for job to be cleared it has no dependencies that
+		 * have errorred. Once we move the job to error it should unschedule all dependant
+		 * jobs recursively.
```

Comment:
Are we sure there's no race conditions that allow this, e.g. a task error just after we decided we needed to call this function. If not, maybe just remove the Assert.

Reply:
It seems like we'd need "repeatable read" isolation level to ensure this. So probably best to remove the check.